### PR TITLE
Multiple sync-related fixes

### PR DESCRIPTION
**Pull request by almet**
I found some issues with the current implementation of the sync mechanism, which this PR fixes.

- The "move a feature from one datalayer to another" feature was not thought of, at all, and as a result introduced the bug that was referenced in #1931.
- It turns out that the syncing of this operation wasn't working either, creating shape clones along the way.
- The previous (buggy) implementation was not sending the proper layer id, resulting to features hosted by the wrong layers.
- Operations were triggered when loading data from the original server, which shouldn't happen.

Generally speaking, the current implementation misses some cross-datalayer tests, but I suggest these should come as a separate pull request.

### CODE REVIEW

File: umap/static/umap/js/umap.features.js [hunk 1]
```diff
@@ -15,21 +15,29 @@ U.FeatureMixin = {
   onCommit: function () {
     // When the layer is a remote layer, we don't want to sync the creation of the
     // points via the websocket, as the other peers will get them themselves.
-    if (this.datalayer.isRemoteLayer()) return
+    if (this.datalayer && this.datalayer.isRemoteLayer()) return
```
Comment (davidbgk) [inline umap/static/umap/js/umap.features.js:18]:
```suggestion
    if (this.datalayer?.isRemoteLayer()) return
```

Does it make a difference I missed?

Comment (almet) [reply umap/static/umap/js/umap.features.js]: Good hint :+1:

File: umap/static/umap/js/umap.features.js [hunk 2]
```diff
@@ -15,21 +15,29 @@ U.FeatureMixin = {
...
+
+    // The "endEdit" event is triggered at the end of an edition,
+    // and will trigger the sync.
+    // In the case of a deletion (or a change of layer), we don't want this
+    // event triggered to cause a sync event, as it would reintroduce
+    // deleted features.
+    // The `._marked_for_deletion` private property is here to track this status.
+    if (this._marked_for_deletion == true) {
+      this._marked_for_deletion == false
```
[the remark this answers is not in the document]
Comment (almet) [reply umap/static/umap/js/umap.features.js]: :upside_down_face: fixed, in [`70a55b6` (#1936)](https://github.com/umap-project/umap/pull/1936/commits/70a55b62d5708ca5f38e2e3dcf2bf0400fc48dbb)

File: umap/static/umap/js/umap.js [hunk 3]
```diff
@@ -190,7 +190,7 @@ U.Map = L.Map.extend({
         dataUrl = decodeURIComponent(dataUrl)
         dataUrl = this.localizeUrl(dataUrl)
         dataUrl = this.proxyUrl(dataUrl)
-        const datalayer = this.createDataLayer()
+        const datalayer = this.createDataLayer({})
```
Comment (yohanboniface) [inline umap/static/umap/js/umap.js:193]: Not sure why this is needed, it seems to me that this parameter already has a default

File: umap/static/umap/js/umap.layer.js [hunk 4]
```diff
@@ -939,11 +939,11 @@ U.DataLayer = L.Evented.extend({
     if (idx !== -1) this._propertiesIndex.splice(idx, 1)
   },
 
-  addData: function (geojson) {
+  addData: function (geojson, sync) {
```
Comment (yohanboniface) [inline umap/static/umap/js/umap.layer.js:942]: Not saying to do it in this PR, but this rings a bell to make sync order out of the internal object process.
In very short, the one who calls `createDataLayer` would also create sync after, or not.

File: umap/static/umap/js/umap.features.js [hunk 5]
```diff
@@ -15,21 +15,29 @@ U.FeatureMixin = {
...
   },
 
-  syncDelete: function () {
-    this.sync.delete()
-  },
-
   initialize: function (map, latlng, options, id) {
     this.map = map
     this.sync = map.sync_engine.proxy(this)
+    this._mark_for_deletion = true
```
Comment (yohanboniface) [inline umap/static/umap/js/umap.features.js:40]: I don't get why this is initialized to `true`.

File: umap/static/umap/js/umap.layer.js (hunk 4 above)
Comment (almet) [reply umap/static/umap/js/umap.layer.js:942]: Correct, that's something to solve at some point. 👍

File: umap/static/umap/js/umap.js (hunk 3 above)
Comment (almet) [reply umap/static/umap/js/umap.js:193]: You're correct. Fixed.

File: umap/static/umap/js/umap.features.js (hunk 5 above)
Comment (almet) [reply umap/static/umap/js/umap.features.js:40]: That's an error. Fixed.

**Comment by almet** [comment]
Rebased on latest master, and incorporated the proposed changed from reviews. We should be ready to go.

File: umap/static/umap/js/umap.js (hunk 3 above)
Comment (davidbgk) [reply umap/static/umap/js/umap.js:193]: Is it back again?

File: umap/static/umap/js/umap.features.js [hunk 6]
```diff
@@ -15,21 +15,29 @@ U.FeatureMixin = {
...
-    if (this.datalayer.isRemoteLayer()) return
+    if (this.datalayer?.isRemoteLayer()) return
+
+    // The "endEdit" event is triggered at the end of an edition,
+    // and will trigger the sync.
+    // In the case of a deletion (or a change of layer), we don't want this
+    // event triggered to cause a sync event, as it would reintroduce
+    // deleted features.
+    // The `._marked_for_deletion` private property is here to track this status.
+    if (this._marked_for_deletion == true) {
```
Comment (davidbgk) [inline umap/static/umap/js/umap.features.js:26]:
```suggestion
    if (this._marked_for_deletion === true) {
```

Is it safer to check for [strict equality](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Equality_comparisons_and_sameness) here?

Comment (almet) [reply umap/static/umap/js/umap.features.js]: yes it is :+1:

**Review by yohanboniface — APPROVED** [review]

**Pull request merged**